Embryonic but functioning JSR223 support

rjolly opened the pull request:
This is mainly to obtain feedback. Once on track, I will consider remaining features like compiled scripts, variable bindings, classpath issues and so on, to be on par with the 2.12 implementation.

How to run:
```
jrunscript -J-Djava.class.path="C:\Program Files\Java\jdk1.8.0_191\lib\tools.jar;dotty-library_0.11-0.11.0-RC1.jar;scala-library-2.12.7.jar;scala-asm-6.0.0-scala-1.jar" -classpath dotty-compiler_2.12-0.12.0-bin-SNAPSHOT-nonbootstrapped.jar;dotty-interfaces-0.11.0-RC1.jar -l scala
scala> "hello world"
val res0: String = hello world

hello world
scala>
```

Blaisorblade commented:
This doesn’t look very invasive *but* it needs tests, unless there’s an extremely good reason.

rjolly commented:
The test would be along the line of:
```
val m = new javax.script.ScriptEngineManager()
val e = m.getEngineByName("scala")
println(e.eval("42"))
```
But first the engine needs to be bootstrapped : we need to have dotty-compiler.jar with /META-INF/services/javax.script.ScriptEngineFactory in the tests classpath. Any suggestion how to do that is welcome.

smarter commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,77 @@
+import java.io.{Reader, StringWriter};
+import javax.script.{AbstractScriptEngine, Bindings, ScriptContext, ScriptEngine => JScriptEngine, ScriptEngineFactory, ScriptException, SimpleBindings}
+
+class ScriptEngine(factory: ScriptEngine.Factory) extends AbstractScriptEngine {
```
Some documentation on what this is, how it works and how to use it would be helpful.

smarter commented on tests/run-with-compiler/scripting.check:
```diff
@@ -0,0 +1,3 @@
+[33mval[0m [36mres0[0m: [35mInt[0m = [31m42[0m
```
Maybe use `-color:off` to avoid having ansi escape codes in test files.

rjolly commented:
@Blaisorblade The problem was with the classloader used by the `new ScriptEngineManager()` call. I fixed it. Then; by putting the test in run-with-compiler, everything works fine.
@smarter I removed the color. The best would be to switch off the output altogether but I do not know how to do that. I added some documentation.

smarter commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+    val vid = state.valIndex
+    state = driver.run(script)(state)
+    val oid = state.objectIndex
+    Class.forName(s"rs$$line$$$oid", true, rendering.classLoader()(state.context))
```
Hardcoding implementation details like the weird classname here isn't great. I haven't looked in details at what this does but could it be achieved by some addition to the API of the REPL instead ?

rjolly replied:
Yes. The class name is computed in Rendering.valueOf(symbol) as called from ReplDriver.displayDefinitions/Members(), themselves called from ReplDriver.compile(). In the ScriptEngine, I am calling ReplDriver.run() which calls interpret() which calls compile(). So in theory, one can decompose these calls to get access to the CompilationUnit and get the Symbol from the tpd.Tree and compute the newestWrapper to get the name.

smarter replied:
Or maybe State could have an additional field for the symbol of the last wrapper module ?

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+    def getLanguageName() = "Scala"
+    def getLanguageVersion() = versionString
+    def getMimeTypes() = asList("application/x-scala")
+    def getNames() = asList("scala")
```
Methods above could be nullary methods

allanrenucci replied:
Prefer `Arrays.asList(...)` to `asList(...)` for the methods above

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+    def getMimeTypes() = asList("application/x-scala")
+    def getNames() = asList("scala")
+
+    def getMethodCallSyntax(obj: String, m: String, args: String*): String = null
```
Why `null`? `s"$obj.$m(${args.mkstring(", ")})"`?

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+
+    def getMethodCallSyntax(obj: String, m: String, args: String*): String = null
+
+    def getOutputStatement(toDisplay: String): String = null
```
s"""print("$toDIsplay")"""?

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+      case _ => null
+    }
+
+    def getProgram(statements: String*): String = null
```
`statements.mkstring("; ")`?

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+
+    def getScriptEngine: JScriptEngine = {
+      new ScriptEngine(this)
+    }
```
No need for curly braces

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+ *
+ *  @param factory the script engine factory
+ */
+class ScriptEngine(factory: ScriptEngine.Factory) extends AbstractScriptEngine {
```
Does it need to take a factory?
```scala
class ScriptEngine extends AbstractScriptEngine {
  ...
  private[this] val driver = new ReplDriver(Array("-usejavacp", "-color:never"), Console.out, None)
  ...
  def getFactory: ScriptEngineFactory = new ScriptEngineFactory {
    ...
    def getScriptEngine = new ScriptEngine
  }
```

rjolly replied:
Well, it avoids to re-instantiate one each time the method is called..

rjolly replied:
Well, it allows to reuse the same driver in all new script engines..

allanrenucci replied:
If you share the same driver, you will share the same classloader across multiple REPL sessions. Is it intended?

rjolly replied:
I have nothing against it.

allanrenucci replied:
To me, this is asking for trouble. If you run two script engines in parallel, they will override each other output and you will get `ClassNotFoundException`...

Unless there is a good reason to share the driver, I would argue against it.

rjolly replied:
ok, done.

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+  import scala.util.Properties.versionString
+
+  class Factory extends ScriptEngineFactory {
+    val driver = new ReplDriver(Array("-usejavacp", "-color:never"), Console.out, None);
```
no semi colon at the end

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+    def getEngineVersion() = "3.0"
+    def getExtensions() = asList("scala")
+    def getLanguageName() = "Scala"
+    def getLanguageVersion() = versionString
```
Prefer `Properties.versionString`

rjolly commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+  @throws[ScriptException]
+  def eval(reader: Reader, context: ScriptContext): Object = eval(stringFromReader(reader), context)
+
+  def stringFromReader(reader: Reader) = {
```
You mean, duplicate the code of eval(String) ?

allanrenucci replied:
No. I meant something like that:
```scala
def eval(reader: Reader, context: ScriptContext): Object = {
  val input = ... // code that was in `stringFromReader`
  eval(input, context)
}
```

rjolly replied:
done.

rjolly commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+  def eval(reader: Reader, context: ScriptContext): Object = eval(stringFromReader(reader), context)
+
+  def stringFromReader(reader: Reader) = {
+    val writer = new StringWriter()
```
Be careful that it is StringWriter, not StringBuffer.

allanrenucci replied:
Yes that's what I don't understand. Why use a `StringWriter` instead of a `StringBuilder`?

rjolly replied:
Well, I take a writer because there is a reader. This is a trick I often use to pipe a reader to a writer. The right way to implement this method would be in a reactive manner, that is evaluate lines as and when they are coming in.

Blaisorblade replied:
> Well, I take a writer because there is a reader. This is a trick I often use to pipe a reader to a writer.

Even then, you can still use `StringBuilder`; the only point of `StringWriter` is if you need something to implement the `Writer` interface, which we don't. Otherwise, `StringBuilder.append(c.toChar)` works.

More importantly, reading *one character at a time* is pretty wasteful — that's what `read(char[] buf)` is for. This waste might be acceptable here.

> The right way to implement this method would be in a reactive manner, that is evaluate lines as and when they are coming in.

Indeed, this loop reads until the end of the stream, that seems problematic as you say. What about testing `reader.ready()` instead?

rjolly replied:
Yes, it might be relevant. However, I would like to point out that this method is seldom used, in Scala or other scripting languages. This is a quick hack to implement the ScriptEngine interface. We could also throw UnsupportedOperationException. If a relevant use case comes, we can always implement it in a clean (and tested) fashion. For the moment, my goal is not to go further than 2.12. What do you think ?

Blaisorblade replied:
> This is a quick hack to implement the ScriptEngine interface. We could also throw UnsupportedOperationException.

👍

allanrenucci commented on compiler/src/dotty/tools/repl/ScriptEngine.scala:
```diff
@@ -0,0 +1,90 @@
+
+object ScriptEngine {
+  import java.util.Arrays.asList
+  import scala.util.Properties.versionString
```
I believe these two imports are now unused

rjolly replied:
fixed.
